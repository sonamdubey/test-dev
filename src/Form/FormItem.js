import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

class FormItem extends React.Component {
  static propTypes = {
    /** The content of the form item. */
    children: PropTypes.node,
    /** The helper message. */
    helperText: PropTypes.string,
    /** A prefix for `FormItem` component classes. */
    prefixClass: PropTypes.string,
    /** The validation status. */
    validateStatus: PropTypes.oneOf(['error', '']),
  };

  static defaultProps = {
    prefixClass: 'oxygen-form-item',
  };

  renderWrapper(child1, child2) {
    const { prefixClass, validateStatus } = this.props;

    let wrapperClassName = `${prefixClass}-control`;

    if (validateStatus) {
      wrapperClassName = classNames(wrapperClassName, {
        'has-error': validateStatus === 'error',
      });
    }

    return (
      <div className={wrapperClassName}>
        {child1}
        {child2}
      </div>
    );
  }

  renderHelper() {
    const { helperText, prefixClass } = this.props;

    if (helperText) {
      const helperTexterClassName = `${prefixClass}-help`;

      return <div className={helperTexterClassName}>{helperText}</div>;
    }

    return null;
  }

  render() {
    const { children, helperText, prefixClass } = this.props;

    let itemClassName = `${prefixClass}`;

    if (helperText) {
      itemClassName = classNames(itemClassName, [`${prefixClass}-with-help`]);
    }

    return (
      <div className={itemClassName}>
        {this.renderWrapper(children, this.renderHelper())}
      </div>
    );
  }
}

export default FormItem;
