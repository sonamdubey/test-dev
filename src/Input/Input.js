import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import omit from 'omit.js';

function fixControlledValue(value) {
  if (typeof value === 'undefined' || value === null) {
    return '';
  }

  return value;
}

class Input extends React.Component {
  static propTypes = {
    /** A custom class for input container. */
    containerClassName: PropTypes.string,
    /** The default value of the `Input` element. */
    defaultValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    /** If `true`, the input will be disabled. */
    disabled: PropTypes.bool,
    /** The `id` of the input element. */
    id: PropTypes.string,
    /** A custom class for input. */
    inputClassName: PropTypes.string,
    /** A custom text for label. */
    label: PropTypes.string,
    /** A custom class for label. */
    labelClassName: PropTypes.string,
    /** Name attribute of the `input` element. */
    name: PropTypes.string,
    /** The short hint displayed in the input before the user enters a value. */
    placeholder: PropTypes.string,
    /** The prefix text/icon for the `Input`. */
    prefix: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
    /** The prefix for component classes. */
    prefixClass: PropTypes.string,
    /** Specify an input field as read-only. */
    readOnly: PropTypes.bool,
    /** Set field as required. */
    required: PropTypes.bool,
    /** The suffix text/icon for the `Input`. */
    suffix: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
    /** Define input type. */
    type: PropTypes.oneOf(['text', 'number', 'tel', 'email', 'password']),
    /** The value of the `input` element. */
    value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    /** Callback fired when input is blurred. */
    onBlur: PropTypes.func,
    /** Callback fired when the value is changed. */
    onChange: PropTypes.func,
    /** Callback fired when input is focused. */
    onFocus: PropTypes.func,
  };

  static defaultProps = {
    prefixClass: 'oxygen-input',
    type: 'text',
  };

  renderAffixInput(children) {
    const { props } = this;

    if (!('prefix' in props || 'suffix' in props)) {
      return children;
    }

    const prefix = props.prefix && (
      <div className={`${props.prefixClass}-prefix`}>{props.prefix}</div>
    );

    const suffix = props.suffix && (
      <div className={`${props.prefixClass}-suffix`}>{props.suffix}</div>
    );

    return (
      <div className={`${props.prefixClass}-affix-content`}>
        {prefix}
        {children}
        {suffix}
      </div>
    );
  }

  render() {
    const { id, label, value, prefixClass } = this.props;

    let { containerClassName, labelClassName, inputClassName } = this.props;

    containerClassName = classNames([`${prefixClass}`], containerClassName);
    labelClassName = classNames([`${prefixClass}-label`], labelClassName);
    inputClassName = classNames([`${prefixClass}-field`], inputClassName);

    // Fix https://fb.me/react-unknown-prop
    const inputProps = omit(this.props, [
      'containerClassName',
      'labelClassName',
      'inputClassName',
      'label',
      'prefix',
      'prefixClass',
      'suffix',
    ]);

    if ('value' in this.props) {
      inputProps.value = fixControlledValue(value);

      // Input elements must be either controlled or uncontrolled,
      // specify either the value prop, or the defaultValue prop, but not both.
      delete inputProps.defaultValue;
    }

    return (
      <div className={containerClassName}>
        {label && (
          <label htmlFor={id} className={labelClassName}>
            {label}
          </label>
        )}
        {this.renderAffixInput(
          <input {...inputProps} className={inputClassName} />,
        )}
      </div>
    );
  }
}

export default Input;
