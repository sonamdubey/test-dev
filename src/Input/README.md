# Input

## Usage
> Important: Make sure to include the [scss file](https://github.com/carwale/oxygen/blob/master/src/Input/style/input.scss) or feel free to create your own.
```js
import { Input } from "oxygen"
// or
import Input from "oxygen/Input"

<Input
  label="Name"
  type="text"
  placeholder="Enter your name"
  value="John Doe"
  onChange={this.handleChange}
/>
```

## API

### Props

| propName | propType | defaultValue | isRequired | description |
| -------- | -------- | ------------ | ---------- | ----------- |
| containerClassName | `string` | - | - | A custom class for input container. |
| defaultValue | oneOf `string`, `number` | - | - | The default value of the `Input` element. |
| disabled | `bool` | - | - |	If `true`, the input will be disabled. |
| id | `string` | - | - | The `id` of the `input` element. |
| inputClassName | `string` | - | - | A custom class for input. |
| label | `string` | - | - | A custom text for label. |
| labelClassName | `string` | - | - | A custom class for label. |
| name | `string` | - | - | Name attribute of the `input` element. |
| placeholder | `string` | - | - | The short hint displayed in the input before the user enters a value. |
| prefix | oneOf `string`, `node` | - | - | The prefix text/icon for the `Input`. |
| prefixClass | `string` | 'oxygen-input' | - | A prefix for `Input` component classes. |
| readOnly | `bool` | - | - | Specify an input field as read-only. |
| required | `bool` | - | - | Set field as required. |
| suffix | oneOf `string`, `node` | - | - | The suffix text/icon for the `Input`. |
| type | oneOf `text`, `number`, `tel`, `email`, `password` | - | Define input type. |
| value | oneOf `string`, `number` | `''` | - | The value of the `input` element. |


### Events

| propName | propType | defaultValue | isRequired | description |
| -------- | -------- | ------------ | ---------- | ----------- |
| onBlur | `func` | - | - | Callback fired when input is blurred. |
| onChange | `func` | - | - | Callback fired when the value is changed. |
| onFocus | `func` | - | - | Callback fired when input is focused. |
