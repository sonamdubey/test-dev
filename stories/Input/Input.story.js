import React from 'react';
import { storiesOf } from '@storybook/react';
import { withInfo } from '@storybook/addon-info';
import { withKnobs, select, text } from '@storybook/addon-knobs';

import Input from '../../src/Input';
import '../../src/Input/style/input.scss';

/* knobs */
// prop: type
const typeLabel = 'type';
const typeOptions = {
  text: 'text',
  tel: 'tel',
  number: 'number',
  email: 'email',
  password: 'password'
};
const typeDefaultValue = 'text';


storiesOf('Input', module)
  .addDecorator(withKnobs) // TODO: make this a global decoration and move to config file [https://storybook.js.org/addons/introduction/]
  .add('default',
    withInfo()(() => (
      <Input
        id="name"
        label={text('label', 'Name')}
        type={select(typeLabel, typeOptions, typeDefaultValue)}
        placeholder={text('placeholder', 'Enter your name')}
      />
    ))
  )
  .add('with affix',
    withInfo()(() => (
      <Input
        id="mobileNumber"
        label={text('label', 'Mobile Number')}
        type="tel"
        placeholder={text('placeholder', 'Enter your mobile number')}
        value={text('value', '9876543210')}
        maxLength="10"
        prefix="+91"
      />
    ))
  )
