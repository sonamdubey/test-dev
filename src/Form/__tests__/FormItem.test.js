import React from 'react';

import FormItem from '../FormItem';
import Input from '../../Input';

describe('<FormItem />', function() {
  describe('render', function() {
    it('should render a form item input', function() {
      const wrapper = shallow(
        <FormItem>
          <Input />
        </FormItem>
      )

      expect(wrapper).toMatchSnapshot();
    })
  })

  describe('props', function() {
    it('should render a form item input', function() {
      const wrapper = mount(
        <FormItem>
          <Input />
        </FormItem>
      )

      expect(wrapper.find('input')).toHaveLength(1);
      expect(wrapper).toMatchSnapshot();
    })

    it('should render a form item with helper text as error', function() {
      const wrapper = shallow(
        <FormItem helperText="Enter a valid name">
          <Input />
        </FormItem>
      )

      //expect(wrapper.find('input')).toHaveLength(1);
      expect(wrapper).toMatchSnapshot();
    })
  })
})
