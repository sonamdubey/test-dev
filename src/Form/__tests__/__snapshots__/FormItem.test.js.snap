// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormItem /> props should render a form item input 1`] = `
<FormItem
  prefixClass="oxygen-form-item"
>
  <div
    className="oxygen-form-item"
  >
    <div
      className="oxygen-form-item-control"
    >
      <Input
        prefixClass="oxygen-input"
        type="text"
      >
        <div
          className="oxygen-input"
        >
          <input
            className="oxygen-input-field"
            type="text"
          />
        </div>
      </Input>
    </div>
  </div>
</FormItem>
`;

exports[`<FormItem /> props should render a form item with helper text as error 1`] = `
<div
  className="oxygen-form-item oxygen-form-item-with-help"
>
  <div
    className="oxygen-form-item-control"
  >
    <Input
      prefixClass="oxygen-input"
      type="text"
    />
    <div
      className="oxygen-form-item-help"
    >
      Enter a valid name
    </div>
  </div>
</div>
`;

exports[`<FormItem /> render should render a form item input 1`] = `
<div
  className="oxygen-form-item"
>
  <div
    className="oxygen-form-item-control"
  >
    <Input
      prefixClass="oxygen-input"
      type="text"
    />
  </div>
</div>
`;
